pkg/v1/layout/write.go

```diff
@@ -127,6 +129,133 @@ func (l Path) AppendDescriptor(desc v1.Descriptor) error {
 	return l.WriteFile("index.json", rawIndex, os.ModePerm)
 }
 
+// ReplaceImage writes a v1.Image to the Path and updates
+// the index.json to reference it, replacing any existing one that matches matcher, if found.
+func (l Path) ReplaceImage(img v1.Image, matcher match.Matcher, options ...Option) error {
+	if err := l.WriteImage(img); err != nil {
+		return err
+	}
+
+	mt, err := img.MediaType()
```

Comment:
Use `partial.Descriptor` here, for some bad reasons and some good reasons.

Reply:
I don't understand

Reply:
Sure, sorry I should have provided more details. It's two things:

1. This is ~9 lines of code to produce a `v1.Descriptor` for an artifact, and we've consolidated that into a `partial.Descriptor` method, just for brevity.
2. There are situations where you have some artifact, say a `v1.Image`, which was originally referenced by a `v1.ImageIndex` descriptor that contained platform information. If you just reconstruct the `v1.Descriptor` as you're doing here, we lose that information. Thus, artifacts can optionally implement `Descriptor() (v1.Descriptor, error)` to return the original `v1.Descriptor` by which they were referenced. This is especially interesting for foreign layers. See [here](https://github.com/google/go-containerregistry/tree/master/pkg/v1/partial#partialdescriptor) and https://github.com/google/go-containerregistry/pull/654.

It unfortunately needs to live in `partial` because the way we wrap things in partial means we shadow the wrapped artifact, but fortunately it makes sense to live there anyway.

Reply:
Oh, I see that I bungled the multiline comment thing in GitHub. This comment should apply to lines 139-158.